### PR TITLE
Draft: Backup codes curtis

### PR DESCRIPTION
**Pull request by rh-iatanaso**

### CODE REVIEW

File: server-spi/src/main/java/org/keycloak/models/credential/BackupCodeCredentialModel.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import org.keycloak.models.credential.dto.BackupCodeSecretData;
+import org.keycloak.util.JsonSerialization;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class BackupCodeCredentialModel extends CredentialModel {
+
+    public static final String TYPE = "backup-code";
```
Comment (rh-iatanaso) [inline server-spi/src/main/java/org/keycloak/models/credential/BackupCodeCredentialModel.java:15]: This should be probably backup-codes or backup-auth-codes ( maybe Andre already changed it)

File: server-spi/src/main/java/org/keycloak/models/utils/BackupAuthnCodesUtils.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
+/*
```
Comment (rh-iatanaso) [inline server-spi/src/main/java/org/keycloak/models/utils/BackupAuthnCodesUtils.java:1]: We should remove the copy right notice it is 2016 . We can ask Marek about copyright notice in the code. But for now I would remove all of them

File: server-spi/src/main/java/org/keycloak/models/utils/BackupAuthnCodesUtils.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * @author <a href="mailto:anascime@redhat.com">Andre Nascimento</a>
+ */
+public class BackupAuthnCodesUtils {
+
+    private static final int NUMBER_OF_CODES = 5;
```
Comment (rh-iatanaso) [inline server-spi/src/main/java/org/keycloak/models/utils/BackupAuthnCodesUtils.java:38]: Let put the required 15

File: server-spi/src/main/java/org/keycloak/models/utils/BackupAuthnCodesUtils.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    private static final RandomString randomString = new RandomString(4, new SecureRandom());
+    public static final int NUM_HASH_ITERATIONS = 1;
+    public static final String NOM_ALGORITHM_TO_HASH = Algorithm.RS256;
```
Comment (rh-iatanaso) [inline server-spi/src/main/java/org/keycloak/models/utils/BackupAuthnCodesUtils.java:41]: I think  RS512 is what Marek wanted

**Comment by rh-iatanaso** [comment]
@cwelchrh @andre-nascimento-rh  We need to add brute force protection.

File: server-spi/src/main/java/org/keycloak/models/credential/BackupCodeCredentialModel.java (hunk 1 above)
Comment (andre-nascimento-rh) [reply server-spi/src/main/java/org/keycloak/models/credential/BackupCodeCredentialModel.java]: Done. I believe I refactored all the assets in this regard to something like: "BackupAuthnCodes..." / "backup-authn-codes...".

File: server-spi/src/main/java/org/keycloak/models/utils/BackupAuthnCodesUtils.java (hunk 2 above)
Comment (andre-nascimento-rh) [reply server-spi/src/main/java/org/keycloak/models/utils/BackupAuthnCodesUtils.java:1]: Done!

File: server-spi/src/main/java/org/keycloak/models/utils/BackupAuthnCodesUtils.java (hunk 3 above)
Comment (andre-nascimento-rh) [reply server-spi/src/main/java/org/keycloak/models/utils/BackupAuthnCodesUtils.java:38]: Done!

Besides that, I also renamed the field name to: "QUANTITY_OF_BACKUP_AUTHN_CODES_TO_GENERATE".

Is that Ok?

**Comment by andre-nascimento-rh** [comment]
> @cwelchrh @andre-nascimento-rh We need to add brute force protection.

Done!

**Pull request closed without merging**